storage, fs: move pebble cluster version compat check to InitEnv

xinhaoz opened the pull request:
### storage: move store version checks into fs pkg
Move functions to perform pebble compatibility checks into the fs pkg.

### storage: add cluster version to env config

### storage, fs: move pebble cluster version compat check to InitEnv

Previously, we performed the CRDB version compatibility check for pebble and the
currently running cluster in the pebble constructor. This commit moves that
check to the file system environment initialization so that we can  colocate
with fs encyrption initialization.

We also extend the version checking to validate that pebble's minversion is not
too high for the current cockroach version, to guard against backwards
incompatible changes.

Fixes: https://github.com/cockroachdb/cockroach/issues/148213

Epic: none
Release note: None

xinhaoz committed:
storage: move store version checks into fs pkg

Move functions to perform pebble compatibility checks into the fs pkg.

xinhaoz committed:
storage: add cluster version to env config

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151935)
<!-- Reviewable:end -->

xinhaoz committed:
storage, fs: move pebble cluster version compatibility check to InitEnv

Previously, we performed the CRDB version compatibility check for pebble and the
currently running cluster in the pebble constructor. This commit moves that
check to the file system environment initialization so that we can  colocate
with fs encyrption initialization.

We also extend the version checking to validate that pebble's minversion is not
too high for the current cockroach version, to guard against backwards
incompatible changes.
Fixes: #148213

Epic: none
Release note: None

jbowens reviewed (commented):
(still reading; flushing comments on first commit)

> Reviewed 10 of 10 files at r1, 1 of 19 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151935#-:-OY252a-3eIBZAEVsd_A:blktp1d)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/storage/minversion/min_version.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151935#-OY24JyCAaQDCeStBRZj:-OY24JyCAaQDCeStBRZk:bmtek3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1465e580d9e2a230963740bbac03cb88bc3395cf/pkg/storage/minversion/min_version.go#L26)):*
> ```Go
> // MinVersionIsAtLeastTargetVersion returns whether the min version recorded
> // on disk is at least the target version.
> func MinVersionIsAtLeastTargetVersion(
> ```

nit: This is moot if we end up moving this into `storage/fs`, but if these are in a `minversion` package I think we should strip the `MinVersion` prefix from these names to avoid the stutter.
___
*[`pkg/storage/min_version.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151935#-OY238EDAwD1buCGLlkb:-OY238EDAwD1buCGLlkc:b-on9vbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1465e580d9e2a230963740bbac03cb88bc3395cf/pkg/storage/min_version.go#L37)):*
> ```Go
> 	}
> 	filename := atomicRenameFS.PathJoin(dir, minversion.MinVersionFilename)
> 	if err := fs.SafeWriteToUnencryptedFile(atomicRenameFS, dir, filename, b, fs.UnspecifiedWriteCategory); err != nil {
> ```

Maybe we should move the min version stuff into the `fs` package itself? it seems like the func to write the min version file (`writeMinVersionFile`) should be in the `minversion` package, but that's not possible because of the cycle that motivated this move right?


<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
Nice

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @jbowens reviewed 25 of 25 files at r4, 8 of 22 files at r5, 2 of 22 files at r6, 15 of 19 files at r7, 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151935#-:-OY6zI4w7j8ElIH9l2tK:bmu9b66)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom)



<!-- Sent from Reviewable.io -->

xinhaoz commented:
tftr!
bors r+
